### ISSUE TITLE
Energy is changing depending on number of OMP_THREADS

### ISSUE THREAD

**Issue by zakmachachi** [dftd4/dftd4#245]
**Describe the bug**
I am running a small test on around 200 atoms and I am finding some bizzare results. When I change the number of OMP_NUM_THREADS in my os.environ I get different energies. For example:

with 28 threads I get: -624.079152864644 (took 1 minute and 8 s to run)
with 14 threads I get: -311.575214051504 (took 58 s to run)
with 1 thread I get: 21.65725415762698 (took 50.3 s to run)

When I run this on my other machine which is not using intel compilers I get:

-21.657254157626973 (12 s to run)

**To Reproduce**
Steps to reproduce the behaviour:

```
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"
os.environ["OMP_NUM_THREADS"] = "1"

from ase.io import read
from dftd4.ase import DFTD4

init_conf = read("GO.xyz")
init_conf.calc = DFTD4(method="pbe")
print(init_conf.get_potential_energy())
```
[GO.txt](https://github.com/dftd4/dftd4/files/15380368/GO.txt)

Change the file to .xyz

I am not sure where to find all the relevant compiler info as I installed this using ASE (conda and then pip as conda was giving me some errors about GOMP 5.0 when I was importing from script)

**Expected behaviour**
Much faster evaluation on machine 1 and not for scaling with OMP

**Comment by marvinfriede** [comment]
I cannot reproduce this behavior. I created a fresh conda environment with ase and dftd4-python. I tried installing both with conda and pip. I always get an energy of -21.657254157626973. The computation time also decreases as expected, i.e., always by the factor of how many OMP threads are used. I did not encounter any issues with OMP. Since the packages from conda and pip (usually) contain precompiled binaries, I do not see a relation to the compilers. Or did you build from source?

**Comment by marvinfriede** [comment]
Are you using Python 3.12? This has not precompiled binaries yet.

Not sure about debugging tools. Debugging a package with a C/Fortran backend is not straightforward. You could check if the Fortran standalone produces correct results. Note, however, that the extended xyz file format is not supported outside of ASE.

Summing up the energies does not exactly add up, but it is somewhat close. Not sure.

**Comment by zakmachachi** [comment]
So the issue appears only when I am using a Jupyter notebook. Really not sure why but this is strange behaviour.. any ideas?

**Comment by marvinfriede** [comment]
> So the issue appears only when I am using a Jupyter notebook. Really not sure why but this is strange behaviour.. any ideas?

Is it reproducible with the jupyter notebook? If so, can you share it?
Are you restarting the kernel for different runs? Maybe some objects persisting in the kernel sessions pose issues.

**Comment by marvinfriede** [comment]
I let Claude have a go at this and it was not able to reproduce the changing energies, but the timing part of the original report.

**TL;DR:** Other OMP runtimes in the environment might hijack D4's and overwrite `OMP_NUM_THREADS`. Torch does this for example, and since the issue was raised by someone who seems to have an ML background, this makes sense. Use `torch.set_num_threads(n)` (or `omp_set_num_threads(n)`) to fix.

Below is the detailed report:

The `dftd4` wheel ships its own OpenMP runtime (`dftd4.libs/libgomp-*.so`), which is loaded `RTLD_LOCAL`. If something that loads an OpenMP runtime `RTLD_GLOBAL` is imported first — PyTorch does — then dftd4's `omp_*` / `GOMP_*` symbols resolve against that runtime instead of its own, and dftd4 silently inherits its thread count. Since `import torch` pins that to 1, `OMP_NUM_THREADS` stops having any effect on dftd4, which is why more threads never got faster.

That matches the timings in the first post (50 s / 58 s / 68 s for 1 / 14 / 28 threads — flat, and slightly worse with more threads).

## Reproducing

```bash
python3 -m venv /tmp/d4repro
/tmp/d4repro/bin/pip install dftd4 ase numpy
/tmp/d4repro/bin/pip install torch --index-url https://download.pytorch.org/whl/cpu
```

Save the script below as `/tmp/d4repro/d4_omp_check.py`, put the `GO.xyz` from this issue next to it, and run:

```bash
cd /tmp/d4repro
./bin/python d4_omp_check.py 8
```

It measures three configurations in one go (torch cannot be un-imported, so it re-executes itself once per configuration) and prints, for each, how many OpenMP runtimes are loaded, how many worker threads were *actually* spawned (counted in `/proc/self/task`) and the speedup *actually* achieved (CPU time / wall time):

```
dftd4 4.2.0   210 atoms   OMP_NUM_THREADS=8

        runtimes max_threads workers speedup    wall   energy / eV
------------------------------------------------------------------
plain          1           8       8    7.6x    7.2s   -21.657254157626973
torch          2           1       1    1.0x   50.1s   -21.65725415762698
fix            2        1->8       8    7.6x    7.0s   -21.657254157626966
```

`plain` is dftd4 on its own, `torch` imports torch first, `fix` additionally calls `omp_set_num_threads()` on the runtime dftd4 ended up on. The thread count is the first argument, and a different structure can be passed as the second.

## Workaround

Use `torch.set_num_threads(n)` (or `omp_set_num_threads(n)` as in the `fix` row). `OMP_NUM_THREADS` cannot control dftd4 once another package owns the OpenMP runtime.

<details>
<summary><code>d4_omp_check.py</code></summary>

```python
#!/usr/bin/env python3
"""Self-contained check for dftd4 issue #245 (OpenMP threading in Python).

One run measures three configurations and prints a comparison:

    plain   dftd4 on its own
    torch   torch imported *before* dftd4 -- torch's libgomp captures dftd4's
            OpenMP symbols, so OMP_NUM_THREADS stops controlling dftd4
    fix     same as torch, plus omp_set_num_threads() on the runtime dftd4
            actually ended up on

For each it reports the OpenMP runtimes loaded, the thread count that runtime
hands out, the OMP workers really spawned (counted in /proc/self/task), and the
parallel speedup really achieved (CPU time / wall time).

Usage:
    python d4_omp_check.py [nthreads] [structure.xyz]

Defaults to 8 threads and ./GO.xyz (the structure attached to the issue).
Requires dftd4, ase and numpy; the torch rows are skipped if torch is missing.
"""

import ctypes
import json
import os
import subprocess
import sys
import threading
import time

MODES = ("plain", "torch", "fix")
DEFAULT_STRUCTURE = "GO.xyz"
DEFAULT_THREADS = 8


def openmp_runtimes():
    """Return every OpenMP runtime mapped into this process."""
    found = set()
    with open("/proc/self/maps") as fh:
        for line in fh:
            path = line.split()[-1]
            if os.path.basename(path).startswith(("libgomp", "libomp", "libiomp")):
                found.add(path)
    return sorted(found)


def omp_handle(runtimes):
    """Return the OpenMP runtime dftd4's calls actually resolve against.

    Symbols are looked up in the process-global scope first, so a runtime that
    some other package loaded RTLD_GLOBAL wins over dftd4's own bundled copy,
    which is loaded RTLD_LOCAL and is therefore invisible globally.
    """
    try:
        lib = ctypes.CDLL(None)
        lib.omp_get_max_threads
        return lib, "shared with another package"
    except AttributeError:
        return ctypes.CDLL(runtimes[0]), "dftd4's own private copy"


def count_threads():
    """Return the number of OS threads currently in this process."""
    return len(os.listdir("/proc/self/task"))


def measure(model, param, hartree):
    """Evaluate the dispersion energy, watching threads and CPU time."""
    baseline = count_threads()
    peak = [baseline]
    done = threading.Event()

    def sample():
        while not done.is_set():
            peak[0] = max(peak[0], count_threads())
            time.sleep(0.002)

    watcher = threading.Thread(target=sample, daemon=True)
    watcher.start()
    wall0, cpu0 = time.perf_counter(), time.process_time()
    energy = model.get_dispersion(param, grad=True)["energy"].item() * hartree
    wall = time.perf_counter() - wall0
    cpu = time.process_time() - cpu0
    done.set()
    watcher.join()

    return energy, peak[0] - baseline, cpu / wall, wall


def run_child(mode, nthreads, path):
    """Measure one configuration; called in a dedicated process."""
    if mode in ("torch", "fix"):
        import torch  # noqa: F401  loaded before dftd4 on purpose

    import dftd4
    import numpy as np
    from ase.io import read
    from ase.units import Bohr, Hartree
    from dftd4.interface import DampingParam, DispersionModel

    runtimes = openmp_runtimes()
    omp, shared = omp_handle(runtimes)
    before = omp.omp_get_max_threads()
    if mode == "fix":
        omp.omp_set_num_threads(nthreads)

    atoms = read(path)
    model = DispersionModel(
        atoms.numbers,
        atoms.positions / Bohr,
        atoms.get_initial_charges().sum(),
        np.array(atoms.cell) / Bohr,
        atoms.pbc,
    )
    energy, workers, speedup, wall = measure(model, DampingParam(method="pbe"), Hartree)

    return {
        "mode": mode,
        "dftd4": dftd4.__version__,
        "natoms": len(atoms),
        "runtimes": runtimes,
        "shared": shared,
        "max_threads": before,
        "max_threads_after": omp.omp_get_max_threads(),
        "workers": workers,
        "speedup": speedup,
        "wall": wall,
        "energy": energy,
    }


def available_modes():
    """Return the configurations that can run in this environment."""
    probe = subprocess.run([sys.executable, "-c", "import torch"], capture_output=True)
    if probe.returncode:
        print("torch not installed -- only the 'plain' configuration will run\n")
        return ("plain",)
    return MODES


def report(results, nthreads):
    """Print the comparison table and the conclusion."""
    first = results[0]
    print(
        f"\ndftd4 {first['dftd4']}   {first['natoms']} atoms   "
        f"OMP_NUM_THREADS={nthreads}\n"
    )
    head = (
        f"{'':7s} {'runtimes':>8s} {'max_threads':>11s} {'workers':>7s} "
        f"{'speedup':>7s} {'wall':>7s}   energy / eV"
    )
    print(head)
    print("-" * len(head))
    for res in results:
        threads = (
            f"{res['max_threads']}->{res['max_threads_after']}"
            if res["mode"] == "fix"
            else str(res["max_threads"])
        )
        print(
            f"{res['mode']:7s} {len(res['runtimes']):8d} {threads:>11s} "
            f"{res['workers']:7d} {res['speedup']:6.1f}x {res['wall']:6.1f}s   "
            f"{res['energy']!r}"
        )

    print(f"\ndftd4's OpenMP runtime was {results[0]['shared']} in 'plain'")
    if len(results) < 2:
        return
    print(f"                       and {results[1]['shared']} in 'torch'")
    for runtime in results[1]["runtimes"]:
        print(f"    {runtime}")
    if results[1]["workers"] <= 1 < results[0]["workers"]:
        print(
            "\n=> reproduced: OMP_NUM_THREADS no longer controls dftd4 once "
            "torch is imported."
        )
        print(
            "   Use torch.set_num_threads(n), or omp_set_num_threads(n) as "
            "in the 'fix' row."
        )


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "--child":
        mode, nthreads, path = sys.argv[2], int(sys.argv[3]), sys.argv[4]
        print(json.dumps(run_child(mode, nthreads, path)))
        return

    args = [arg for arg in sys.argv[1:] if not arg.startswith("-")]
    nthreads = int(args[0]) if args else DEFAULT_THREADS
    path = args[1] if len(args) > 1 else DEFAULT_STRUCTURE
    if not os.path.exists(path):
        sys.exit(
            f"structure {path!r} not found -- pass one as the second argument, "
            "or download GO.xyz from the issue"
        )

    env = dict(os.environ, OMP_NUM_THREADS=str(nthreads))
    results = []
    for mode in available_modes():
        print(f"running {mode!r} ...", flush=True)
        out = subprocess.run(
            [
                sys.executable,
                os.path.abspath(__file__),
                "--child",
                mode,
                str(nthreads),
                path,
            ],
            env=env,
            capture_output=True,
            text=True,
        )
        if out.returncode:
            print(out.stderr.strip()[-2000:])
            sys.exit(f"configuration {mode!r} failed")
        results.append(json.loads(out.stdout.strip().splitlines()[-1]))

    report(results, nthreads)


if __name__ == "__main__":
    main()
```

</details>
